WIP: change lowering of ccall cconvert arguments

vtjnash opened the pull request:
`ccall(:a, B, (C,), d)` is now lowered as:

```
  d#1 = cconvert(C, d)
  Expr(:call, :ccall, :a, B, (C,), convert(C, d#1), d#1)
```

This allows implementation of the special-case ccall lowering logic in Julia.
It also means that the fall-back cconvert method is now a no-op, with the main
work being done by convert. But it also means we get a GC root for both
the result of cconvert and convert (as needed).

ps: this was the motivation for https://github.com/JuliaLang/julia/pull/9729

simonbyrne committed:
fix rational vs float comparisons (attempt 2)

simonbyrne committed:
fix float/rational vs mathconst comparisons

simonbyrne committed:
add NEWS item

vtjnash committed:
eliminate pointertoref Intrinsic and just use jl_value_ptr function reversably

samuela committed:
Add mapreduce for DArrays

skariel committed:
Factorization is now type stable

Previously retured types depended on the boolean value of a keyword argument `pivot`. Now the api relies on dispatch on different tyeps either `pivot=Val{true}` or `pivot=Val{false}`. A few more type instabilities were mitigated by using a `copy_oftype` method.

pfitzseb committed:
fix numeric instability of signif

and added tests for two of those cases

pfitzseb committed:
two more tests

pfitzseb committed:
use specialized exponention

eschnett committed:
Add more consistency tests when checking `fma`

JeffBezanson committed:
juliatypes.jl: start fixing tests for diagonals

[ci skip]

JeffBezanson committed:
Merge branch 'master' of github.com:JuliaLang/julia

twadleigh committed:
Added stub for workflow tips in the manual.

JeffBezanson committed:
Merge branch 'master' of github.com:JuliaLang/julia

vtjnash committed:
change lowering of ccall cconvert arguments

ccall(:a, B, (C,), d) is now lowered as:
  d#1 = cconvert_gcroot(C, d)
  Expr(:call, :ccall, :a, B, (C,), cconvert(C, d#1), d#1)

This allows implementation of the special-case ccall lowering logic in Julia
and still create a gcroot for it as needed

the Ref{T} type is the new supertype of Ptr{T}. the ccall emission
turns a Ref{T} into a Ptr{T}, as if the &addressOf operator had been
applied. the specific rules for this will be moved from the current
gist to the documentation at a later point.

JeffBezanson commented on base/pcre.jl:
```diff
@@ -100,6 +101,7 @@ end
 function study(regex::Ptr{Void}, options::Integer)
     # NOTE: options should always be zero in current PCRE
     errstr = Array(Ptr{UInt8},1)
+    errstr[1] = C_NULL
```
Looks like a bug fix? Could this go directly to master?

vtjnash replied:
mostly this was caused because I accidentally was passing a copy of the array in the pointer, rather than the array itself (during work on this PR). it shouldn't strictly be necessary on master (since pcre should set this variable)

JeffBezanson commented:
Really nice change. The amount of code deletion speaks for itself :)

JeffBezanson commented on src/ccall.cpp:
```diff
@@ -370,62 +250,71 @@ static Value *julia_to_native(Type *ty, jl_value_t *jt, Value *jv,
+        *needStackRestore = true;
+        Value *jvt = emit_typeof(jv);
+        Value *p;
+        BasicBlock *mutableBB = BasicBlock::Create(getGlobalContext(),"is-mutable",ctx->f);
```
This is a great hack. It looks correct to me; hope I'm right about that :)

vtjnash replied:
(added WIP title)

yeah, i'll definately have to do more testing on this. i probably will hold off on merging until I get a few more of the ccall-related PR's ready. the runtime code branch is identical to the compile-time code branch directly above it, however.

vtjnash replied:
oops, it /was/ slightly broken.

vtjnash commented:
> Really nice change. The amount of code deletion speaks for itself :)

indeed. i'm also working on deprecating `&` (#8134 and https://github.com/JuliaLang/julia/compare/jn/add_ref_type and https://gist.github.com/vtjnash/7296634) which should delete some more of this code.

JeffBezanson commented:
Could we remove the extra `d#1` argument from `ccall`? It looks like it will always be rooted anyway. In fact I'm having trouble remembering why it is there in the current system.

vtjnash commented:
it's possibly there to help inference correctly count usages correctly? i'm not really sure

JeffBezanson commented:
Does the travis failure make any sense to you?

If neither of us can figure out why the extra root is there we should try removing it.

vtjnash commented:
yeah, that should be fixed in my followup fix (two bugs: i was looking for `mutabl` at the wrong offset in the type (rounded to the nearest `i8*` offset, and forgot to add the special case for `Ptr{Void}&`

andreasnoack committed:
Make diag work for AbstractMatrix and a test.

andreasnoack committed:
Fix #9915. When solving SparseMatrixCSC\SparseMatrixCSC and the rhs is symmetric,
the rhs shouldn't be stored with symmetric flag set.

samuela committed:
Add tests for DArray mapreduce

MikeInnes committed:
docs.jl wasn't taking up enough horizontal space

MikeInnes committed:
basic markdown equality

MikeInnes committed:
handle empty paragraph

MikeInnes committed:
basic markdown tests

MikeInnes committed:
flesh out md tests

MikeInnes committed:
cleanup

MikeInnes committed:
improve terminal formatting resets

MikeInnes committed:
hard line breaks

closes #9410

dhoegh committed:
Added a test to cover a untested code path and changed the color in strings to use with_output_color, instead of hardcoded ANSI escape code colors.

MikeInnes committed:
tweaks and tests

twadleigh committed:
Mention option to wrap tests in module.

carnaval committed:
Added missing write barriers in full compilation mode. Should fix #9923.

timholy committed:
Merge pull request #9922 from dhoegh/test_for_#9485

Added a test to cover a untested code path.

samuela committed:
Avoid floating point errors in test

samuela committed:
actually fix this time

andreasnoack committed:
Merge pull request #9813 from samuela/darray_mapreduce

Add mapreduce for DArrays

ViralBShah committed:
Merge pull request #9910 from twadleigh/workflow-tips

Added stub for workflow tips in the manual.

amitmurthy committed:
closes #9841. [ci skip]

amitmurthy committed:
Merge pull request #9929 from amitmurthy/amitm/shdocfix

closes #9841. [ci skip]

simonster committed:
Improve test coverage for scale/scale!

Many of these methods were already covered at least in part via calls
from other functions, but I figure they should be tested explicitly as
well.

carnaval committed:
cleanup gc_verify

carnaval committed:
repair weak references

carnaval committed:
Merge pull request #9931 from JuliaLang/ob/gcfix

Fix weak references

MikeInnes committed:
add include docs

fixes #9454

MikeInnes committed:
fix code quoting

tkelman committed:
Fix conflicting import of Markdown.term warning

when running tests in serial

JeffBezanson committed:
Merge branch 'master' of github.com:JuliaLang/julia

JeffBezanson committed:
fix #9904, hang in front end on nd comprehension

JeffBezanson committed:
fix #9866, `edit(func)`

- error and ask for types if there is > 1 method
- otherwise edit the only existing method

JeffBezanson committed:
fix #9897, missing error fallbacks for promoting operators

JeffBezanson committed:
fix #9789, sign bit is not checked for unequal-size signed/unsigned conversions

vtjnash committed:
keep up with llvm-svm changes

MikeInnes committed:
treat whitespace better in help

pfitzseb committed:
better handling of large digits and digits=0

simonbyrne committed:
Merge pull request #9880 from Varanas/master

Fix numeric instability in signif

Ismael-VC committed:
Document multiline anonymous function.

Version using `function` keyword with no name.
Correcting a typo.

simonbyrne committed:
some more Float16 methods

ivarne committed:
Merge pull request #9937 from Ismael-VC/patch-1

[ci skip] Document multiline anonymous function with `function ... end`.

JeffBezanson committed:
fix #9686, better message for throwing a type instead of an instance in test_throws

JeffBezanson committed:
Merge pull request #9895 from eschnett/more-fma-tests

Add more consistency tests when checking `fma`

JeffBezanson committed:
Merge pull request #9940 from simonbyrne/float16funcs

some more Float16 methods

MikeInnes committed:
basic fuzzy searching in help mode

MikeInnes committed:
whitespace again

MikeInnes committed:
spelling correction

MikeInnes committed:
more selective spellcheck

JeffBezanson committed:
improve accuracy of --track-allocation by visiting the previous line the
next time we hit a branch. this catches lines at the end of a loop.

timholy committed:
Merge pull request #9946 from JuliaLang/jb/trackallocbranches

improve accuracy of --track-allocation

MikeInnes committed:
algorithmic tweaks

MikeInnes committed:
small refactor

MikeInnes committed:
Merge branch 'omm/fuzzy'

andreasnoack committed:
Merge pull request #9575 from skariel/typestable_qrfact

WIP: typestable factorizations

StefanKarpinski committed:
doc: use the phrase "do-block" in the title section on do-blocks

[ci skip]

JeffBezanson committed:
Fix #9947

amitmurthy committed:
fix bug in worker-to-worker connection setup. closes #9951

MikeInnes committed:
fix keyword docs / edge case

amitmurthy committed:
Merge pull request #9953 from amitmurthy/amitm/9951

fix bug in worker-to-worker connection setup. closes #9951

amitmurthy committed:
parallel test fixup [ci skip]

amitmurthy committed:
Merge pull request #9954 from amitmurthy/amitm/testfix

parallel test fixup [ci skip]

amitmurthy committed:
remove unwarranted handling of SIGPIPE instead of ignoring it.

StefanKarpinski committed:
Merge pull request #9198 from simonbyrne/rational3

fix rational vs float comparisons (attempt 2)

vtjnash committed:
Merge branch 'jn/gensym2'

Conflicts:
	src/julia-syntax.scm

vtjnash committed:
Merge pull request #9956 from amitmurthy/amitm/sigpipe

remove unwarranted handling of SIGPIPE instead of ignoring it.

vtjnash committed:
Merge pull request #9952 from JuliaLang/kms/fix9947

Fix #9947

vtjnash committed:
codegen: insert unreachable blocks after throwing errors, rather than Br

vtjnash committed:
Merge remote-tracking branch 'origin/master'

vtjnash committed:
fix #7779 (code-coverage flags with MCJIT)

vtjnash committed:
hide the JIT-function generated names from the global symbol table in sys.so

ihnorton committed:
Add Base.set_process_title, Base.get_process_title

Not exported. Fixes #9957.

vtjnash committed:
fix possible merge error with jn/gensym2

dhoegh committed:
Fix so the `show_method_candidates` do not fail for method with 0 args.

ivarne committed:
Merge pull request #9968 from dhoegh/#8651

Fix `show_method_candidates` so it do not fail for methods with zero args.

amitmurthy committed:
Inbuilt managers throw on unused arguments. Closes #9958

amitmurthy committed:
Merge pull request #9971 from amitmurthy/amitm/9958

Inbuilt managers throw on unused arguments. Closes #9958

jiahao committed:
Docs: fix formatting error in qrfact

Ref: f9da849d

ReST tables must have not have text in between columns delimited by headers.

eschnett committed:
Correct rounding mode after calling llvm_fma intrinsic

Keno committed:
custom libc++ fixes I had lying around

Keno committed:
LLVM 3.7 moved the DebugInfo headers

amitmurthy committed:
more fixes for worker-worker connection setups

amitmurthy committed:
Merge pull request #9979 from amitmurthy/amitm/9951_morefixes

more fixes for worker-worker connection setups

simonbyrne committed:
Merge pull request #9925 from eschnett/correct-fma

Correct rounding mode after calling fma
